### ISSUE TITLE
Potential regression in DAG creation (5.26.1 <-> 5.27.4)

### ISSUE THREAD

**Issue by SnakeBugReporter** [snakemake/snakemake#796]
Hi,

after updating my Conda environment I noticed, that my Snakemake workflow was not working anymore. By testing various versions of the conda package "snakemake-minimal" I was able to narrow things down to two versions of Snakemake, being:
5.26.1 -> workflow is still working
5.27.4 -> workflow is no longer working

My own workflow is about the evaluation of next generation sequencing data (involving fastqc, bwa mem, various GATK applications). The problem is, that the "fastqc rule" gets executed twice and in doing so the workflow entirely fails.

Here is a part of the job overview:
```
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 24
Rules claiming more threads will be scaled down.
Job counts:
        count   jobs
        1       all
        1       bgzip_reference
        2       calculate_average_coverage
        1       create_bwa_index
        4       fastqc
        2       gather_intervals
        2       index_marked_bam
        2       mapping
        2       mark_duplicates
        1       scatter_intervals
        2       sort_sam_gatk
```
Each of the four fastqc rules get executed twice (fastqc itself is executed successfully), e.g.:
```
[Wed Dec  9 16:52:14 2020]
rule fastqc:
    input: FASTQ_INPUT_DIR/SRR2131199_R2.fastq.gz
    output: 00_FastQC/SRR2131199_R2_fastqc.html, 00_FastQC/SRR2131199_R2_fastqc.zip
    log: logs/00_FastQC/SRR2131199_R2.log
    jobid: 4
    benchmark: benchmarks/00_FastQC/SRR2131199_R2.bm
    wildcards: sample=SRR2131199_R2

... more output here ...

[Wed Dec  9 16:52:14 2020]
rule fastqc:
    input: FASTQ_INPUT_DIR/SRR2131199_R2.fastq.gz
    output: 00_FastQC/SRR2131199_R2_fastqc.html, 00_FastQC/SRR2131199_R2_fastqc.zip
    log: logs/00_FastQC/SRR2131199_R2.log
    jobid: 4
    benchmark: benchmarks/00_FastQC/SRR2131199_R2.bm
    wildcards: sample=SRR2131199_R2
```
As soon as the rule finishes there is an error message:
```
Error in rule fastqc:
    jobid: 4
    output: 00_FastQC/SRR2131199_R2_fastqc.html, 00_FastQC/SRR2131199_R2_fastqc.zip
    log: logs/00_FastQC/SRR2131199_R2.log (check log file(s) for error message)
    shell:
        export _JAVA_OPTIONS='-Xmx1g -XX:ParallelGCThreads=4'
fastqc -o 00_FastQC -f fastq FASTQ_INPUT_DIR/SRR2131199_R2.fastq.gz 2> logs/00_FastQC/SRR2131199_R2.log
        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)

Removing output files of failed job fastqc since they might be corrupted:
00_FastQC/SRR2131199_R2_fastqc.html, 00_FastQC/SRR2131199_R2_fastqc.zip
exception calling callback for <Future at 0x7f99387694c0 state=finished raised RuleException>
Traceback (most recent call last):
  File "/usr/local/envs/myWorkflow/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2264, in run_wr
apper
    run(
  File "/home/ubuntu/Volume/test_OV_intervals/Snakefile", line 455, in __rule_fastqc
    threads:
  File "/usr/local/envs/myWorkflow/lib/python3.8/site-packages/snakemake/shell.py", line 202, in __new__
    del cls._processes[jobid]
KeyError: 4

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/envs/myWorkflow/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 560, in _callba
ck
    raise ex
  File "/usr/local/envs/myWorkflow/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/envs/myWorkflow/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 546, in cached_
or_run
    run_func(*args)
  File "/usr/local/envs/myWorkflow/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2349, in run_wr
apper
    raise RuleException(
snakemake.exceptions.RuleException: KeyError in line 434 of /home/ubuntu/Volume/test_OV_intervals/Snakefile:
4
  File "/usr/local/envs/myWorkflow/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2264, in run_wr
apper
  File "/home/ubuntu/Volume/test_OV_intervals/Snakefile", line 434, in __rule_fastqc

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/envs/myWorkflow/lib/python3.8/concurrent/futures/_base.py", line 328, in _invoke_callbacks
    callback(self)
  File "/usr/local/envs/myWorkflow/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 572, in _callba
ck
    error_callback(job)
  File "/usr/local/envs/myWorkflow/lib/python3.8/site-packages/snakemake/scheduler.py", line 583, in _error
    self._handle_error(job)
  File "/usr/local/envs/myWorkflow/lib/python3.8/site-packages/snakemake/scheduler.py", line 593, in _handle_error
    self.running.remove(job)
KeyError: fastqc
```

After various tests, I **guess** that the problem is related to DAG reevaluation after a checkpoint. I tested various target rules and the problem only occurs for rules after the checkpoint.

I also created a minimalistic example of my workflow, which I want to share:
```
FASTQ_R1 = ["GGA081_R1", "GGA112_R1"]
FASTQ_R2 = ["GGA081_R2", "GGA112_R2"]
FASTQ    = ["GGA081", "GGA112"]
FastQ_in_path="FASTQ"

rule all:
    input:
        expand("00_FastQC/{sample}_fastqc.html", sample=FASTQ_R1 + FASTQ_R2),
        expand("00_FastQC/{sample}_fastqc.zip", sample=FASTQ_R1 + FASTQ_R2),
        expand("03_combined/{sample}.sam", sample=FASTQ)
        #expand("02_split/{sample}", sample=FASTQ)
        #expand("01_mapping/{sample}.sam", sample=FASTQ)

rule fastqc:
    input:
        # sample shall not be expanded here => {{sample}}
        expand("{path}/{{sample}}.fastq.gz", path=FastQ_in_path)
    output:
        "00_FastQC/{sample}_fastqc.html",
        "00_FastQC/{sample}_fastqc.zip"
    log:
        "logs/00_FastQC/{sample}.log"
    shell:
        # -Xmx1g is needed for small files, GC=4 just to be save
        "export _JAVA_OPTIONS='-Xmx1g -XX:ParallelGCThreads=4'\n"
        "fastqc -o 00_FastQC -f fastq {input} 2> {log}"

rule mapping:
    input:
        fq1=expand("{path}/{{sample}}_R1.fastq.gz", path=FastQ_in_path),
        fq2=expand("{path}/{{sample}}_R2.fastq.gz", path=FastQ_in_path),
        genome="processed_reference/SampleSeq.fa.gz",
        bwaIndex="processed_reference/SampleSeq.fa.gz.bwt"
    output:
        "01_mapping/{sample}.sam"
    threads:
        6
    resources:
        mem_gb=4
    log:
        bwa_err="logs/01_mapping/{sample}_stderr_bwa.log",
        samtools_err="logs/01_mapping/{sample}_stdout_samtools.log"
    shell:
        "bwa mem -M -t {threads} "
        "{input.genome} {input.fq1} {input.fq2} 2> {log.bwa_err} > {output}"

heckpoint scatter:
    input:
        sam="01_mapping/{sample}.sam"
    output:
        pieces=directory("02_split/{sample}")
    shell:
        "mkdir 02_split/{wildcards.sample}\n"
        "split -l 500000 {input.sam} 02_split/{wildcards.sample}/PIECE"

def aggregate_input(wildcards):
    outDir = checkpoints.scatter.get(**wildcards).output[0]
    return expand("02_split/{{sample}}/PIECE{suffix}", suffix=glob_wildcards(outDir + "/PIECE{match}").match)

rule gather:
    input:
        sams=aggregate_input
    output:
        "03_combined/{sample}.sam"
    shell:
        "cat {input.sams} > {output}"
```
This workflow is running flawlessly with Snakemake 5.26.1 but fails with 5.27.4. Strangely here the rule ``fastqc`` doesn't cause the problem, but the same mapping job is executed twice. Thereby the workflow fails again. So basically the same problem with a different tool/rule.

I hope this helps you to improve Snakemake even further.

Thanks a lot!

Edit:
The directory structure with the input data for the sample workflow looks as follows:
```
FASTQ/
  GGA081_R1.fastq.gz  GGA112_R1.fastq.gz
  GGA081_R2.fastq.gz  GGA112_R2.fastq.gz
processed_reference/
  SampleSeq.fa.gz      SampleSeq.fa.gz.ann  SampleSeq.fa.gz.pac
  SampleSeq.fa.gz.amb  SampleSeq.fa.gz.bwt  SampleSeq.fa.gz.sa
```
Fastqs are rather small (approx. 50-60 MB).

**Comment by ASLeonard** [comment]
This is the same issue as reported in #727, which is still without a resolution.

**Comment by SnakeBugReporter** [comment]
Thanks ASLeonard! You're right seems to be the same issue.

**Comment by pb-cdunn** [comment]
This is fixed by `5.30.2`. You can close this.

**Comment by dlaehnemann** [comment]
I think that PR #808 finally fixed this (alongside issues #806 and #727), also see this comment:
https://github.com/snakemake/snakemake/issues/806#issuecomment-747393210

So I'll close it for now, but feel free to reopen, if this persists.
